File: en/codes/kotlin/chapter_hashing/array_hash_map.kt
/**
 * File: array_hash_map.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package chapter_hashing

/* Key-value pair */
class Pair(
    var key: Int,
    var _val: String
)

/* Hash table based on array implementation */
class ArrayHashMap {
    // Initialize an array, containing 100 buckets
    private val buckets = arrayOfNulls<Pair>(100)

    /* Hash function */
    fun hashFunc(key: Int): Int {
        val index = key % 100
        return index
    }

    /* Query operation */
    fun get(key: Int): String? {
        val index = hashFunc(key)
        val pair = buckets[index] ?: return null
        return pair._val
    }

    /* Add operation */
    fun put(key: Int, _val: String) {
        val pair = Pair(key, _val)
        val index = hashFunc(key)
        buckets[index] = pair
    }

    /* Remove operation */
    fun remove(key: Int) {
        val index = hashFunc(key)
        // Set to null, indicating removal
        buckets[index] = null
    }

    /* Get all key-value pairs */
    fun pairSet(): MutableList<Pair> {
        val pairSet = mutableListOf<Pair>()
        for (pair in buckets) {
            if (pair != null)
                pairSet.add(pair)
        }
        return pairSet
    }

    /* Get all keys */
    fun keySet(): MutableList<Int> {
        val keySet = mutableListOf<Int>()
        for (pair in buckets) {
            if (pair != null)
                keySet.add(pair.key)
        }
        return keySet
    }

    /* Get all values */
    fun valueSet(): MutableList<String> {
        val valueSet = mutableListOf<String>()
        for (pair in buckets) {
            if (pair != null)
                valueSet.add(pair._val)
        }
        return valueSet
    }

    /* Print hash table */
    fun print() {
        for (kv in pairSet()) {
            val key = kv.key
            val _val = kv._val
            println("$key -> $_val")
        }
    }
}

/* Driver Code */
fun main() {
    /* Initialize hash table */
    val map = ArrayHashMap()

    /* Add operation */
    // Add key-value pair (key, value) to the hash table
    map.put(12836, "Ha")
    map.put(15937, "Luo")
    map.put(16750, "Suan")
    map.put(13276, "Fa")
    map.put(10583, "Ya")
    println("\nAfter adding, the hash table is\nKey -> Value")
    map.print()

    /* Query operation */
    // Enter key to the hash table, get value
    val name = map.get(15937)
    println("\nInput student ID 15937, found name $name")

    /* Remove operation */
    // Remove key-value pair (key, value) from the hash table
    map.remove(10583)
    println("\nAfter removing 10583, the hash table is\nKey -> Value")
    map.print()

    /* Traverse hash table */
    println("\nTraverse key-value pairs Key -> Value")
    for (kv in map.pairSet()) {
        println("${kv.key} -> ${kv._val}")
    }
    println("\nIndividually traverse keys Key")
    for (key in map.keySet()) {
        println(key)
    }
    println("\nIndividually traverse values Value")
    for (_val in map.valueSet()) {
        println(_val)
    }
}

File: en/codes/typescript/chapter_hashing/array_hash_map.ts
/**
 * File: array_hash_map.ts
 * Created Time: 2022-12-20
 * Author: Daniel (better.sunjian@gmail.com)
 */

/* Key-value pair Number -> String */
class Pair {
    public key: number;
    public val: string;

    constructor(key: number, val: string) {
        this.key = key;
        this.val = val;
    }
}

/* Hash table based on array implementation */
class ArrayHashMap {
    private readonly buckets: (Pair | null)[];

    constructor() {
        // Initialize an array, containing 100 buckets
        this.buckets = new Array(100).fill(null);
    }

    /* Hash function */
    private hashFunc(key: number): number {
        return key % 100;
    }

    /* Query operation */
    public get(key: number): string | null {
        let index = this.hashFunc(key);
        let pair = this.buckets[index];
        if (pair === null) return null;
        return pair.val;
    }

    /* Add operation */
    public set(key: number, val: string) {
        let index = this.hashFunc(key);
        this.buckets[index] = new Pair(key, val);
    }

    /* Remove operation */
    public delete(key: number) {
        let index = this.hashFunc(key);
        // Set to null, indicating removal
        this.buckets[index] = null;
    }

    /* Get all key-value pairs */
    public entries(): (Pair | null)[] {
        let arr: (Pair | null)[] = [];
        for (let i = 0; i < this.buckets.length; i++) {
            if (this.buckets[i]) {
                arr.push(this.buckets[i]);
            }
        }
        return arr;
    }

    /* Get all keys */
    public keys(): (number | undefined)[] {
        let arr: (number | undefined)[] = [];
        for (let i = 0; i < this.buckets.length; i++) {
            if (this.buckets[i]) {
                arr.push(this.buckets[i].key);
            }
        }
        return arr;
    }

    /* Get all values */
    public values(): (string | undefined)[] {
        let arr: (string | undefined)[] = [];
        for (let i = 0; i < this.buckets.length; i++) {
            if (this.buckets[i]) {
                arr.push(this.buckets[i].val);
            }
        }
        return arr;
    }

    /* Print hash table */
    public print() {
        let pairSet = this.entries();
        for (const pair of pairSet) {
            console.info(`${pair.key} -> ${pair.val}`);
        }
    }
}

/* Driver Code */
/* Initialize hash table */
const map = new ArrayHashMap();
/* Add operation */
// Add key-value pair (key, value) to the hash table
map.set(12836, 'Ha');
map.set(15937, 'Luo');
map.set(16750, 'Suan');
map.set(13276, 'Fa');
map.set(10583, 'Ya');
console.info('\nAfter adding, the hash table is\nKey -> Value');
map.print();

/* Query operation */
// Enter key to the hash table, get value
let name = map.get(15937);
console.info('\nEnter student ID 15937, found name ' + name);

/* Remove operation */
// Remove key-value pair (key, value) from the hash table
map.delete(10583);
console.info('\nAfter removing 10583, the hash table is\nKey -> Value');
map.print();

/* Traverse hash table */
console.info('\nTraverse key-value pairs Key->Value');
for (const pair of map.entries()) {
    if (!pair) continue;
    console.info(pair.key + ' -> ' + pair.val);
}
console.info('\nIndividually traverse keys Key');
for (const key of map.keys()) {
    console.info(key);
}
console.info('\nIndividually traverse values Value');
for (const val of map.values()) {
    console.info(val);
}

export {};
